### PR TITLE
Resolve all TODO v5.0

### PR DESCRIPTION
**Pull request by aslesarenko**
This PR do code fixes marked with `// TODO v5.0` and the following changes
- [x] fix CostingBox.equals
- [x] moved safeNewArray to common module
- [x] added VersionContext
- [x] simplified `verify()` and `prove()` methods 
- [x] VersionContext.isJitActivated method added
- [x] throw ValidationException for Option in DataSerializer
- [x] extend CheckSerializableTypeCode validation rule to cover OptionTypeCode 
- [x] tests of ErgoTrees which build deeply nested collections
- [x] unify usage of CheckPositionLimit to allow v6.0 soft-fork
- [x] ensure all Colls are limited by MaxArrayLength
- [x] checkSoftForkCondition both in Interpreter and ProverInterpreter

### CODE REVIEW

File: sigmastate/src/main/scala/sigmastate/serialization/ValueSerializer.scala [hunk 1]
```diff
@@ -347,6 +347,7 @@ object ValueSerializer extends SigmaSerializerCompanion[Value[SType]] {
     }
   }
 
+  // TODO mainnet v5.0: control maxTreeDepth same as in deserialize
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/serialization/ValueSerializer.scala:350]: Do GH issue, not todo

File: sigmastate/src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala [hunk 2]
```diff
@@ -189,8 +189,8 @@ object ErgoBoxCandidate {
       val tree = DefaultSerializer.deserializeErgoTree(r, SigmaSerializer.MaxPropositionSize)  // READ
       val creationHeight = r.getUInt().toInt    // READ
       val nTokens = r.getUByte()                // READ
-      val tokenIds = ValueSerializer.newArray[Array[Byte]](nTokens)
-      val tokenAmounts = ValueSerializer.newArray[Long](nTokens)
+      val tokenIds = safeNewArray[Array[Byte]](nTokens)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala:192]: Yes, but I think it is better to have consistent usage of the same allocation method in all serializer (where the actual size is deserialized). The overhead is negligible.
I suggest to leave it as-is.

File: sigmastate/src/main/scala/sigmastate/eval/CostingDataContext.scala [hunk 3]
```diff
@@ -350,10 +349,15 @@ case class CostingBox(isCost: Boolean, val ebox: ErgoBox) extends Box with Wrapp
 
   override def hashCode(): Int = Ints.fromByteArray(id.toArray)
 
-  override def equals(obj: Any): Boolean = (this eq obj.asInstanceOf[AnyRef]) || (obj != null && ( obj match {
-    case obj: Box => util.Arrays.equals(id.toArray, obj.id.toArray)
-    // TODO v5.0: return false when obj in not Box instead of throwing an exception
-  }))
+  override def equals(obj: Any): Boolean = (this eq obj.asInstanceOf[AnyRef]) || (obj != null && {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/CostingDataContext.scala:352]: When EQ or NEQ operation is deserialized the arguments should be of the same type, which is checked in the node builder as part of deserialization here https://github.com/ScorexFoundation/sigmastate-interpreter/blob/5531dce6778e46a8663d4c8ecbf6bfe998710b41/sigmastate/src/main/scala/sigmastate/lang/SigmaBuilder.scala#L739
So attack is not possible by crafting malicious script.

File: sigmastate/src/main/scala/org/ergoplatform/validation/ValidationRules.scala [hunk 4]
```diff
@@ -204,13 +204,24 @@ object ValidationRules {
...
+    }
+
     final def apply[T](typeCode: Byte): Unit = {
       checkRule()
       val ucode = toUByte(typeCode)
-      if (ucode > toUByte(OpCodes.LastDataType)) {
-        throwValidationException(
-          new SerializerException(s"Data value of the type with the code $ucode cannot be deserialized."),
-          Array(typeCode))
+      if (typeCode == SOption.OptionTypeCode || ucode > toUByte(OpCodes.LastDataType)) {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/validation/ValidationRules.scala:218]: Clarified in the code comment.

File: sigmastate/src/main/scala/org/ergoplatform/validation/ValidationRules.scala [hunk 5]
```diff
@@ -309,10 +320,32 @@ object ValidationRules {
...
 
-  /** This rule doesn't have it's own validation logic, however it is used in creation of
-    * ValidationExceptions, which in turn can be checked for soft-fork condition using `this.isSoftFork`. */
+  /** This rule it is used in creation of ValidationExceptions, which in turn can be
+    * checked for soft-fork condition using `this.isSoftFork`. Thus, this rule can be
+    * replaced with a new rule and the limit can be increased.
+    */
   object CheckPositionLimit extends ValidationRule(1014,
     "Check that the Reader has not exceeded the position limit.") with SoftForkWhenReplaced {
+
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/validation/ValidationRules.scala:332]: This rule is introduced to make checkPositionLimit method soft-forkable. Clarified the change in ScalaDoc https://github.com/ScorexFoundation/sigmastate-interpreter/blob/a1f62481ea0f09cff6a1bf5c8bd19a203f62d3a5/sigmastate/src/main/scala/sigmastate/utils/SigmaByteReader.scala#L37

File: sigmastate/src/main/scala/sigmastate/serialization/ValueSerializer.scala (hunk 1 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/serialization/ValueSerializer.scala:350]: added issue https://github.com/ScorexFoundation/sigmastate-interpreter/issues/781

**Review by kushti — APPROVED** [review]
LGTM node, but please add notes that complexity is not accounted in the cost in JIT mode anymore,this change is important

**Pull request merged**